torch/csrc/jit/tensorexpr/bounds_overlap.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <torch/csrc/jit/tensorexpr/ir_simplifier.h>
+#include <torch/csrc/jit/tensorexpr/ir_visitor.h>
+#include <torch/csrc/jit/tensorexpr/stmt.h>
+#include <deque>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+namespace bounds {
+
+// A simple class containing the start and end of a range in a single dimension.
+struct TORCH_API Bound {
+  const Expr* start{nullptr};
+  const Expr* end{nullptr};
+
+  // This stores whether or not the start and end of this Bound have previously
+  // been swapped. This occurs when the bound is in a loop with a negative
+  // stride.
+  bool swapped{false};
+
+  Bound() = default;
+  Bound(const Expr* s, const Expr* e) : start(s), end(e) {}
+
+  void print() const {
+    std::cout << "(" << *start << ", " << *end << ")";
+  }
+
+  bool equals(const Bound& other) const {
+    return exprEquals(start, other.start) && exprEquals(end, other.end);
+  }
+
+  void swap() {
+    std::swap(start, end);
+    swapped = !swapped;
+  }
+};
+
+enum OverlapKind {
+  NoOverlap = 0,
+  PartialOverlap,
+  TotalOverlapA,
+  TotalOverlapB
+};
+
+// Returns the kind of overlap between Bound a and Bound b in a single
+// dimension.
+//     PartialOverlap if any elements in the bound b are in the bound a.
+//     TotalOverlapA if all elements in the bound b are in the bound a.
+//     TotalOverlapB if all elements in the bound a are in the bound b, or if
+//        they are equal.
```

Comment:
Why not have another category if they are equal and fully overlapping? Having `TotalOverlapB` represent this case seems arbitrary and sometimes confusing while reading through the code.

Reply:
Honestly, it's because the overlap code is used in the context of subtraction and theres no difference in those cases. I'd rather not complicate it further unless we need the distinction.